server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/sse/S3SSECustomerKeyGenerator.java

```diff
@@ -0,0 +1,54 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.objectstorage.aws.sse;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Base64;
+
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+import com.google.common.hash.Hashing;
+
+public class S3SSECustomerKeyGenerator {
+
+    public static final int ITERATION_COUNT = 65536;
+    public static final int KEY_LENGTH = 256;
+    public static final SecretKeyFactory SECRET_KEY_FACTORY;
+
+    static {
+        try {
+            SECRET_KEY_FACTORY = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Could not initialize SecretKeyFactory", e);
+        }
+    }
+
+    public static String generateCustomerKey(String masterPassword, String salt) throws InvalidKeySpecException {
+        PBEKeySpec spec = new PBEKeySpec(masterPassword.toCharArray(), salt.getBytes(), ITERATION_COUNT, KEY_LENGTH);
+        byte[] derivedKey = SECRET_KEY_FACTORY.generateSecret(spec).getEncoded();
```

Comment:
Can this code be called from multiple threads? Because SecretKeyFactory#generateSecret is not pure nor thread-safe.

Reply:
> Can this code be called from multiple threads?

yes

> Because SecretKeyFactory#generateSecret is not pure and it does not guarantee thread-safety AFAIK.

I just write concurrent test case for it
see commit https://github.com/apache/james-project/pull/2488/commits/73fd56a0b807663bea924c24fce49d6db588edcc

It works well.

Reply:
Sorry for force push,
the test case for it: `S3SSECustomerKeyGeneratorTest#generateShouldWorkCorrectWhenConcurrent`